### DISCUSSION TITLE
MS Teams

### DISCUSSION THREAD

**Discussion by rzuritamilla** [johnste/finicky#227] [General]
Hi, 

Perhaps a(nother) newbie question but the [MS Teams handler](https://github.com/johnste/finicky/wiki/Configuration-ideas#open-microsoft-teams-links-in-the-native-app) does not work as expected. Finicky does open MS Teams but it does not go to the corresponding url (I am clicking a link that should take me to a file). Instead MS Teams stays where I last left it (file or team or chat window). Do you have any tip for me?

**Comment by johnste** [comment]
Sorry, I'm not a user of MS Teams myself, so I won't be able to help you here.

**Comment by thomasLeclaire** [comment]
Hello, 
just got a look and yes this ms teams handler does not work as expected. 

There is several things : 

- first the teams internal url is not msteams://xxxx but msteams:/xxxwith only ONE slash. (the microsoft magic..)
so it need to have a custom version here for teams. I just provided a PR for this purpose: https://github.com/johnste/finicky/pull/169

- Also, the common links shared in invitations (at least in my company!) have structure like this one : 
https://teams.microsoft.com/dl/launcher/launcher.html?url=%2F_%23%2Fl%2Fmeetup-join%2F19%3Ameeting_xxxx&type=meetup-join&deeplinkId=uselessHere&directDl=true&msLaunch=true&enableMobilePage=true&suppressPrompt=true

which need to be converted into : 
msteams:/l/meetup-join/19:meeting_xxx  with xxx url-decoded
all the arguments are globally useless. 

ie these rewrites : 
```
    {
      // handle teams redirects
      match: /teams.microsoft.com\/l/,
      url:  ({urlString}) => decodeURIComponent(urlString.match(/^.*teams.microsoft.com\/(.*?)\?.*/i)[1])
    },
    {
      // handle teams redirects
      match: /teams.microsoft.com\/dl/,
      url:  ({url}) => (decodeURIComponent(url.search.match(/^url=.*(l%2F.*?)&type/i)[1]).match(/^(.*)\?.*/)[1])
    }
```
    
   and this handler : 
```
    {
      match: /l\/meetup-join|l\/channel/,
      browser: 'com.microsoft.teams',
      url({ url }) {
        return {
          ...url,
          protocol: 'msteams',
        };
      },
    },   
```

**Comment by gerwitz** [comment]
I have success with a very simple handler:
```
    {
      match: /^https?:\/\/teams\.microsoft\.com\/l\/meetup\-join\/.*$/,
      browser: "Microsoft Teams"
    }
```

**Comment by thomasLeclaire** [comment]
> I have success with a very simple handler:
> 
> ```
>     {
>       match: /^https?:\/\/teams\.microsoft\.com\/l\/meetup\-join\/.*$/,
>       browser: "Microsoft Teams"
>     }
> ```

not working on my side, like there is still the double slash issue. 

which Teams version do you have?  and do you have any example of a full link? 

by the way, your handler match only one type of teams links, but there are the webcast, the files, the channel ones at minima.

**Comment by gerwitz** [comment]
Teams version 1.4.00.11161, but this has worked for quite some time over many updates.

 Indeed I have only one use case, and it is for joining scheduled events. I’m in a big corporate IT setting, and created this to handle the “join” links that appear in calendar invites that sometimes aren’t recognized by Teams to trigger a nice join button (and before Outlook even tried.

**Comment by santry** [comment]
I'm on Teams version 1.5.00.31156 and macOS Ventura 13.0.1 and the other solutions in this thread have not worked for me. I came up with the following, which _does_ work for launching meetings for me:

```javascript
module.exports = {
  defaultBrowser: "Safari",

  rewrite: [{
    match: ({ url }) => url.host === "teams.microsoft.com",
    url({ url }) {
      // Build the URL string manually so we get the required single "/" between "msteams:" and "l/meetup-join…"
      return "msteams:" + decodeURI(url.pathname) + "?" + url.search
    }
  }],

  handlers: [{
      match: ({ url }) => url.protocol === "msteams",
      browser: "Microsoft Teams"
  }]
}
```

**Reply by adamcstephens** [reply]
Thanks! This works perfectly.

**Comment by michael-odell** [comment]
This convo is old, but I found it today.  And between the discussion and some experiments, I found something that seems to work for me.  Maybe someone else will find it useful, too.

On a mac with Sonomo 14.6.1 and Finicky v3.4.0 and MS Teams 24277.3502.3161.3007, the following handler is working for me with Teams-generated links to messages, channels, and meetings:

```
        {
            match: ({ url }) => url.host === "teams.microsoft.com",
            browser: "Microsoft Teams"
        }
```

_However_, the links I'm seeing in meeting invites from Teams look different than the original poster's -- they begin with `https://teams.microsoft.com/l/meetup-join/`.  I'm not sure whether that's a universal change or unique to my environment.
